ldap/servers/slapd/conntable.c

```diff
@@ -118,40 +136,60 @@ Connection *
 connection_table_get_connection(Connection_Table *ct, int sd)
 {
     Connection *c = NULL;
-    int index, count;
+    size_t index = 0;
+    size_t count = 0;
+
+    PR_Lock(ct->table_mutex);
+
+    /*
+     * We attempt to loop over the ct twice, because connection_is_free uses trylock
+     * and some resources *might* free in the time needed to loop around.
+     */
+    size_t ct_max_loops = ct->size * 2;
 
+    /*
+     * This uses sd as entropy to randomly start inside the ct rather than
+     * always head-loading the list. Not my idea, but it's probably okay ...
+     */
     index = sd % ct->size;
-    for (count = 0; count < ct->size; count++, index = (index + 1) % ct->size) {
+
+    for (count = 0; count < ct_max_loops; count++, index = (index + 1) % ct->size) {
         /* Do not use slot 0, slot 0 is head of the list of active connections */
         if (index == 0) {
             continue;
-        } else if (ct->c[index].c_mutex == NULL) {
+        } else if (ct->c[index].c_state == CONN_STATE_FREE) {
             break;
-        }
-
-        if (connection_is_free(&(ct->c[index]), 1 /*use lock */)) {
+        } else if (connection_is_free(&(ct->c[index]), 1 /*use lock */)) {
             break;
         }
     }
 
-    if (count < ct->size) {
+    /* If count exceeds max loops, we didn't find something into index. */
+    if (count < ct_max_loops) {
         /* Found an available Connection */
         c = &(ct->c[index]);
         PR_ASSERT(c->c_next == NULL);
         PR_ASSERT(c->c_prev == NULL);
         PR_ASSERT(c->c_extension == NULL);
-        if (c->c_mutex == NULL) {
-            PR_Lock(ct->table_mutex);
-            c->c_mutex = PR_NewMonitor();
-            c->c_pdumutex = PR_NewLock();
-            PR_Unlock(ct->table_mutex);
-            if (c->c_mutex == NULL || c->c_pdumutex == NULL) {
-                c->c_mutex = NULL;
-                c->c_pdumutex = NULL;
-                slapi_log_err(SLAPI_LOG_ERR, "connection_table_get_connection", "PR_NewLock failed\n");
-                exit(1);
-            }
-        }
+        
+	if(c->c_state == CONN_STATE_INIT) {
```

Comment:
I am not sure this block is valid.
c_mutex has been allocated during connection_table_new and c_state=CONN_STATE_INIT.

So we should realloc c_mutex only if CONN_STATE_FREE. Shouldn't the test be c->c_state==CONN_STATE_FREE ?

Also, once the connection table has been allocated, all slots should be CONN_STATE_INIT and c_mutex allocated. When is it necessary to reinit (alloc c_mutex and c_pdumutex) again ?

Reply:
> 

@tbordaz 
Yes, you are correct in both counts. We dont need to realloc c_mutex at all, it was some legacy stuff that I forgot to remove...  Thanks